pkg/recovery/etcd.go

```diff
@@ -0,0 +1,116 @@
+// The etcd backend fetches control plane objects directly from etcd. This is adapted heavily from
+// kubernetes/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go.
+
+package recovery
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/pkg/api"
+)
+
+// etcdBackend is a backend that extracts a controlPlane from an etcd instance.
+type etcdBackend struct {
+	client     *clientv3.Client
+	decoder    runtime.Decoder
+	pathPrefix string
+}
+
+// NewEtcdBackend constructs a new etcdBackend for the given client and pathPrefix.
+func NewEtcdBackend(client *clientv3.Client, pathPrefix string) Backend {
+	return &etcdBackend{
+		client:     client,
+		decoder:    api.Codecs.UniversalDecoder(),
+		pathPrefix: pathPrefix,
+	}
+}
+
+// read implements Backend.read().
+func (s *etcdBackend) read(ctx context.Context) (*controlPlane, error) {
+	cp := &controlPlane{}
+	for _, r := range []struct {
+		etcdKeyName string
+		yamlName    string
+		obj         runtime.Object
+	}{{
+		etcdKeyName: "configmaps",
+		yamlName:    "config-map",
+		obj:         &cp.configMaps,
+	}, {
+		etcdKeyName: "daemonsets",
+		yamlName:    "daemonset",
+		obj:         &cp.daemonSets,
+	}, {
+		etcdKeyName: "deployments",
+		yamlName:    "deployment",
+		obj:         &cp.deployments,
+	}, {
+		etcdKeyName: "poddisruptionbudgets",
+		yamlName:    "pod-disruption-budget",
+		obj:         &cp.podDisruptionBudgets,
```

Comment:
This can be dropped, we wouldn't need this to recover a control-plane (just an added protection from accidentally destroying one - but they should still exist in etcd anyway).